authz/audit_logger.go

```diff
@@ -0,0 +1,111 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package authz
+
+import (
+	"context"
+	"sync"
+)
+
+var (
+	mu sync.Mutex
+	// m is a map from the name to the audit logger builder.
+	// This backs the RegisterAuditLoggerBuilder API.
+	m = make(map[string]AuditLoggerBuilder)
+)
+
+// RegisterAuditLoggerBuilder registers the builder in a global map
+// using b.Name() as the key.
+// This should only be called during initialization time (i.e. in an init() function).
+// If multiple builders are registered with the same name, the one registered last
+// will take effect.
+func RegisterAuditLoggerBuilder(b AuditLoggerBuilder) {
+	mu.Lock()
+	defer mu.Unlock()
+	m[b.Name()] = b
+}
+
+// AuditInfo contains information used by the audit logger during an audit logging event.
+type AuditInfo struct {
+	// RPCMethod is the method of the audited RPC, in the format of "/pkg.Service/Method".
+	// For example, "/helloworld.Greeter/SayHello".
+	RPCMethod string `json:"rpc_method,omitempty"`
```

Comment:
Is this JSON stuff here because we need it internally in order to make any audit logger work, or because we're planning to have one specific implementation of an audit logger that wants to output the data as JSON?  If the latter, then I don't think we want this here.

Reply:
Good point. I removed these annotations.